### PR TITLE
[Connectors API] Unify timestamp field parsing

### PR DESCRIPTION
**Pull request by jedrazb**
## Changes
- Unify `datetime` parsing logic with `parseInstant` function applied to all datetime fields from user in connectors APIs - user can provide datetime in any ES-compatible format, the API will take care of making it compatible with [connectors protocol](https://github.com/elastic/connectors/blob/main/docs/CONNECTOR_PROTOCOL.md)

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java [hunk 1]
```diff
@@ -105,18 +105,24 @@ public ConnectorSyncInfo(StreamInput in) throws IOException {
...
+            LAST_ACCESS_CONTROL_SYNC_SCHEDULED_AT_FIELD.getPreferredName(),
+            ConnectorUtils.formatInstantToFrameworkString(lastAccessControlSyncScheduledAt)
+        );
         if (lastDeletedDocumentCount != null) {
             builder.field(LAST_DELETED_DOCUMENT_COUNT_FIELD.getPreferredName(), lastDeletedDocumentCount);
         }
-        builder.field(LAST_INCREMENTAL_SYNC_SCHEDULED_AT_FIELD.getPreferredName(), lastIncrementalSyncScheduledAt);
+        builder.field(
+            LAST_INCREMENTAL_SYNC_SCHEDULED_AT_FIELD.getPreferredName(),
+            ConnectorUtils.formatInstantToFrameworkString(lastIncrementalSyncScheduledAt)
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:117]: As I'm seeing now `formatInstantToFrameworkString` I'm wondering, if we should fix it the other way around. As the Connectors API guards everything around the protocol it feels cleaner, if the clients (like the framework) adapt to the API and not the other way around. In theory the connectors API should also be useable for other clients/schedulers.

**Review by timgrein — APPROVED** [review]
From a functional perspective it looks correct, but after rethinking about the solution I'm not 100% sure, if the fix should happen inside the API, but rather in the client (connectors framework). Explained my rationale in the comment. 

Still, feel free to merge it, if you think this solution is the right path forward 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java (hunk 1 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:117]: I agree that the issue should be fixed at the framework level, and I already filed an enhancement request https://github.com/elastic/connectors/issues/2067 to unify handling (parsing, representing) datetimes in connectors framework. 

While this PR doesn't solve the underlying root cause it mitigates the symptoms and allows us move forward with integration. Also, think about the connectors created in older releases, we need to be backward compatible with their representation with timestamps (or prepare a migration). Once the issue is solved in framework we can change the API formatting of timestamps.

**Review by navarone-feekery — APPROVED** [review]
LGTM, nice one!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java (hunk 1 above)
Comment (timgrein) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:117]: Thanks for the explanation, makes sense!

**Comment by jedrazb** [comment]
@timgrein @navarone-feekery After discussing this in the team sync, I'm adapting this PR to just address the issue of datetime parsing. Since it's not a bugfix I'm removing the label and not backporting this.

**Review by timgrein — APPROVED** [review]
LGTM, thank you for addressing the changes. Looks way cleaner to remove the framework specific logic out of the API 👍

**Comment by jedrazb** [comment]
This change was actually fixing a bug in `8.12` that caused intermittent build failures due to unit tests failing. The culprit is `Instant.ofEpochMilli(randomLong())` that is changed in this PR to `randomInstant()` that produces instants in reasonable datetime ranges. This is a fix for:
- https://github.com/elastic/elasticsearch/issues/105863
- https://github.com/elastic/elasticsearch/issues/106356

This PR is not adding a feature, it's unifying the logic we use to handle timestamps for connectors, therefore I'm backporting this to 8.12 to address the build failures.

**Comment by jedrazb** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.12|[](https://github.com/elastic/elasticsearch/pull/107045)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**